File: hbase-client/src/main/java/org/apache/hadoop/hbase/security/SaslUtil.java
/**
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.hadoop.hbase.security;

import java.util.Base64;
import java.util.Map;
import java.util.TreeMap;

import javax.security.sasl.Sasl;
import javax.security.sasl.SaslClient;
import javax.security.sasl.SaslException;
import javax.security.sasl.SaslServer;

import org.apache.hadoop.hbase.util.Bytes;
import org.apache.yetus.audience.InterfaceAudience;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

@InterfaceAudience.Private
public class SaslUtil {
  private static final Logger LOG = LoggerFactory.getLogger(SaslUtil.class);
  public static final String SASL_DEFAULT_REALM = "default";
  public static final int SWITCH_TO_SIMPLE_AUTH = -88;

  public enum QualityOfProtection {
    AUTHENTICATION("auth"),
    INTEGRITY("auth-int"),
    PRIVACY("auth-conf");

    private final String saslQop;

    QualityOfProtection(String saslQop) {
      this.saslQop = saslQop;
    }

    public String getSaslQop() {
      return saslQop;
    }

    public boolean matches(String stringQop) {
      if (saslQop.equals(stringQop)) {
        LOG.warn("Use authentication/integrity/privacy as value for rpc protection "
            + "configurations instead of auth/auth-int/auth-conf.");
        return true;
      }
      return name().equalsIgnoreCase(stringQop);
    }
  }

  /** Splitting fully qualified Kerberos name into parts */
  public static String[] splitKerberosName(String fullName) {
    return fullName.split("[/@]");
  }

  public static String encodeIdentifier(byte[] identifier) {
    return Base64.getEncoder().encodeToString(identifier);
  }

  public static byte[] decodeIdentifier(String identifier) {
    return Base64.getDecoder().decode(Bytes.toBytes(identifier));
  }

  public static char[] encodePassword(byte[] password) {
    return Base64.getEncoder().encodeToString(password).toCharArray();
  }

  /**
   * Returns {@link org.apache.hadoop.hbase.security.SaslUtil.QualityOfProtection}
   * corresponding to the given {@code stringQop} value.
   * @throws IllegalArgumentException If stringQop doesn't match any QOP.
   */
  public static QualityOfProtection getQop(String stringQop) {
    for (QualityOfProtection qop : QualityOfProtection.values()) {
      if (qop.matches(stringQop)) {
        return qop;
      }
    }
    throw new IllegalArgumentException("Invalid qop: " +  stringQop
        + ". It must be one of 'authentication', 'integrity', 'privacy'.");
  }

  /**
   * @param rpcProtection Value of 'hbase.rpc.protection' configuration.
   * @return Map with values for SASL properties.
   */
  public static Map<String, String> initSaslProperties(String rpcProtection) {
    String saslQop;
    if (rpcProtection.isEmpty()) {
      saslQop = QualityOfProtection.AUTHENTICATION.getSaslQop();
    } else {
      String[] qops = rpcProtection.split(",");
      StringBuilder saslQopBuilder = new StringBuilder();
      for (int i = 0; i < qops.length; ++i) {
        QualityOfProtection qop = getQop(qops[i]);
        saslQopBuilder.append(",").append(qop.getSaslQop());
      }
      saslQop = saslQopBuilder.substring(1);  // remove first ','
    }
    Map<String, String> saslProps = new TreeMap<>();
    saslProps.put(Sasl.QOP, saslQop);
    saslProps.put(Sasl.SERVER_AUTH, "true");
    return saslProps;
  }

  static void safeDispose(SaslClient saslClient) {
    try {
      saslClient.dispose();
    } catch (SaslException e) {
      LOG.error("Error disposing of SASL client", e);
    }
  }

  static void safeDispose(SaslServer saslServer) {
    try {
      saslServer.dispose();
    } catch (SaslException e) {
      LOG.error("Error disposing of SASL server", e);
    }
  }
}

File: connection/sasl-util.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
#include "connection/sasl-util.h"

#include <glog/logging.h>
#include <sasl/sasl.h>
#include <sasl/saslplug.h>
#include <sasl/saslutil.h>

#include <string>

int SaslUtil::GetPluginPath(void *context __attribute__((unused)), const char **path) {
  *path = getenv("SASL_PATH");

  if (*path == NULL) {
    *path = kDefaultPluginDir;
  }
  return SASL_OK;
}

void *SaslUtil::MutexNew(void) {
  auto m = new std::mutex();
  return m;
}

int SaslUtil::MutexLock(void *m) {
  (reinterpret_cast<std::mutex *>(m))->lock();
  return SASL_OK;
}

int SaslUtil::MutexUnlock(void *m) {
  (reinterpret_cast<std::mutex *>(m))->unlock();
  return SASL_OK;
}

void SaslUtil::MutexDispose(void *m) {
  std::mutex *mutex = reinterpret_cast<std::mutex *>(m);
  delete mutex;
}

std::once_flag SaslUtil::library_inited_;

void SaslUtil::InitializeSaslLib() {
  std::call_once(library_inited_, []() {
    sasl_set_mutex(reinterpret_cast<sasl_mutex_alloc_t *>(&SaslUtil::MutexNew),
                   reinterpret_cast<sasl_mutex_lock_t *>(&SaslUtil::MutexLock),
                   reinterpret_cast<sasl_mutex_unlock_t *>(&SaslUtil::MutexUnlock),
                   reinterpret_cast<sasl_mutex_free_t *>(&SaslUtil::MutexDispose));
    static sasl_callback_t callbacks[] = {
        {SASL_CB_GETPATH, (sasl_callback_ft)&SaslUtil::GetPluginPath, NULL},
        {SASL_CB_LIST_END, NULL, NULL}};
    int rc = sasl_client_init(callbacks);
    if (rc != SASL_OK) {
      throw std::runtime_error("Cannot initialize client " + std::to_string(rc));
    }
  });
}

std::string SaslUtil::ParseServiceName(std::shared_ptr<hbase::Configuration> conf, bool secure) {
  if (!secure) {
    return std::string();
  }
  std::string svrPrincipal = conf->Get(kServerPrincipalConfKey, "");
  // principal is of this form: hbase/23a03935850c@EXAMPLE.COM
  // where 23a03935850c is the host (optional)
  std::size_t pos = svrPrincipal.find("/");
  if (pos == std::string::npos && svrPrincipal.find("@") != std::string::npos) {
    pos = svrPrincipal.find("@");
  }
  if (pos == std::string::npos) {
    throw std::runtime_error("Couldn't retrieve service principal from conf");
  }
  VLOG(1) << "pos " << pos << " " << svrPrincipal;
  std::string service_name = svrPrincipal.substr(0, pos);
  return service_name;
}
